compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3484,26 +3504,54 @@ class Typer extends Namer
-          report.error(em"the result of an implicit conversion must be more specific than $pt", tree.srcPos)
-        inferView(tree, pt) match {
+      def searchTree(t: Tree)(fail: SearchFailure => Tree)(using Context) = {
+        inferView(t, pt) match {
           case SearchSuccess(found: ExtMethodApply, _, _) =>
             found // nothing to check or adapt for extension method applications
           case SearchSuccess(found, _, _) =>
             checkImplicitConversionUseOK(found)
             withoutMode(Mode.ImplicitsEnabled)(readapt(found))
           case failure: SearchFailure =>
-            if (pt.isInstanceOf[ProtoType] && !failure.isAmbiguous) then
-              // don't report the failure but return the tree unchanged. This
-              // will cause a failure at the next level out, which usually gives
-              // a better error message. To compensate, store the encountered failure
-              // as an attachment, so that it can be reported later as an addendum.
-              rememberSearchFailure(tree, failure)
-              tree
-            else recover(failure.reason)
+            fail(failure)
         }
-      else recover(NoMatchingImplicits)
+      }
+
+      val treeTpe = tree.tpe
+
+      def tryUnsafeNullConver(fail: => Tree)(using Context): Tree =
+        // If explicitNulls and unsafeNulls are enabled, and
+        if ctx.mode.is(Mode.UnsafeNullConversion) && pt.isValueType &&
+          treeTpe.isUnsafelyConvertable(pt)
+        then tree.cast(pt)
+        else fail
+
+      def cannotFind(failure: SearchFailure) =
+        if (pt.isInstanceOf[ProtoType] && !failure.isAmbiguous) then
+          // don't report the failure but return the tree unchanged. This
+          // will cause a failure at the next level out, which usually gives
+          // a better error message. To compensate, store the encountered failure
+          // as an attachment, so that it can be reported later as an addendum.
+          rememberSearchFailure(tree, failure)
+          tree
+        else recover(failure.reason)
+
+      val searchCtx =
+        if ctx.explicitNulls && config.Feature.enabled(nme.unsafeNulls) then
+          ctx.addMode(Mode.UnsafeNullConversion)
+        else ctx
+
+      inContext(searchCtx) {
+        if ctx.mode.is(Mode.ImplicitsEnabled) && tree.typeOpt.isValueType then
+          if pt.isRef(defn.AnyValClass) || pt.isRef(defn.ObjectClass) then
+            // We want to allow `null` to `AnyRef` if UnsafeNullConversion is enabled
+            if !(ctx.mode.is(Mode.UnsafeNullConversion) &&
+              treeTpe.isUnsafelyConvertable(pt)) then
+              report.error(em"the result of an implicit conversion must be more specific than $pt", tree.srcPos)
+            tree.cast(pt)
+          else
+            searchTree(tree)(failure => tryUnsafeNullConver(cannotFind(failure)))
+        else tryUnsafeNullConver(recover(NoMatchingImplicits))
+      }
```

Comment:
This looks like it odes what we want. But it makes code that is already complex a lot more complicated. Are there ways to streamline this? I am not after adding more stuff, but maybe we can cleverly re-arrange the logic so it becomes clearer.